usaspending_api/bulk_download/v2/views.py

```diff
@@ -0,0 +1,313 @@
+        except TypeError:
+            raise InvalidParameterException('award_types parameter not provided as a list')
+        # if the filter is calling everything, just remove the filter, save on the query performance
+        if set(award_types) != set(itertools.chain(*award_type_mappings.values())):
+            and_queryset_filters['{}__in'.format(value_mappings[award_level]['type'])] = award_types
+
+        # Adding date range filters
+        # Get the date type attribute
+        date_attribute = None
+        if filters['date_type'] == 'action_date':
+            date_attribute = value_mappings[award_level]['action_date']
+        elif filters['date_type'] == 'last_modified_date':
+            date_attribute = value_mappings[award_level]['last_modified_date']
+        else:
+            raise InvalidParameterException('Invalid parameter for date_type: {}'.format(filters['date_type']))
+        # Get the date ranges
+        try:
+            if 'start_date' in filters['date_range']:
+                and_queryset_filters['{}__gte'.format(date_attribute)] = filters['date_range']['start_date']
+            if 'end_date' in filters['date_range']:
+                and_queryset_filters['{}__lte'.format(date_attribute)] = filters['date_range']['end_date']
+        except TypeError:
+            raise InvalidParameterException('date_range parameter not provided as an object')
+
+        # Agencies are to be OR'd together and then AND'd to the major query
+        or_queryset = (Q(awarding_agency_id=filters['agency']) |
+                       Q(funding_agency_id=filters['agency']))
+        if 'sub_agency' in filters:
+            or_queryset = (or_queryset & (Q(awarding_agency_id=filters['sub_agency']) |
+                                          Q(funding_agency_id=filters['sub_agency'])))
+
+        # Put it all together
+        table = value_mappings[award_level]['table']
+        if and_queryset_filters is not None and or_queryset is not None:
+            filtered_queryset = table.objects.filter(**and_queryset_filters).filter(or_queryset)
+        elif and_queryset_filters is not None:
+            filtered_queryset = table.objects.filter(**and_queryset_filters)
+        elif or_queryset is not None:
+            filtered_queryset = table.objects.filter(or_queryset)
+        else:
+            filtered_queryset = table.objects.all()
+
+        # Note: Ordering by date for testing/verification for now
+        #       If there's a request to sort by a specific attribute,
+        #       here's where it will be.
+        order_by = date_attribute if date_attribute else value_mappings[award_level]['action_date']
+        filtered_queryset = filtered_queryset.order_by(order_by)
+
+        return filtered_queryset
+
+    def get_csv_sources(self, json_request):
+        """
+        Generate the CSV Sources (filtered queryset and metadata) based on the request
+        """
+        award_levels = json_request['award_levels']
```

Comment:
I've seen `.get()` used in the codebase to prevent python key errors if there are missing post data

Reply:
I check to see if the json_request has these fields earlier (see lines 94-96) but I see this definitely applies for the 'columns' and pushing that.